### PR TITLE
added regex based gsub to remove trailing ':'

### LINKED ISSUE #14768: msfconsole: Using a module with a trailing ':' matches all modules
This seems like a bug (albeit minor).

Using a module with with a trailing `:` lists info for all modules instead of selecting the module. Dumping 4000+ lines of info to console is not pleasant. This also makes no sense in instances where the preceding text is clearly a valid module name/path, and cannot possibly match all 4000+ results.

This behavior seems to be inconsistent as some characters at the end of the module name are ignored (`.`) and some are rejected (`*`), whereas `:` appears to be special.

Presumably the `:` character is being used to check if the input is a search filter rather than a module path, and when this value is followed by nothing the filter is presumed to match everything.

I will never need or desire the `use` command to dump info for all modules.

```
msf6 > use exploit/linux/local/nested_namespace_idmap_limit_priv_esc;
[-] No results from search
[-] Failed to load module: exploit/linux/local/nested_namespace_idmap_limit_priv_esc;
```

```
msf6 > use exploit/linux/local/nested_namespace_idmap_limit_priv_esc*
[-] No results from search
[-] Failed to load module: exploit/linux/local/nested_namespace_idmap_limit_priv_esc*
```

```
msf6 > use exploit/linux/local/nested_namespace_idmap_limit_priv_esczzz
[-] No results from search
[-] Failed to load module: exploit/linux/local/nested_namespace_idmap_limit_priv_esczzz
```

```
msf6 > use exploit/linux/local/nested_namespace_idmap_limit_priv_esc|
[-] No results from search
[-] Failed to load module: exploit/linux/local/nested_namespace_idmap_limit_priv_esc|
```

```
msf6 > use exploit/linux/local/nested_namespace_idmap_limit_priv_esc.
[*] Using configured payload linux/x86/meterpreter/reverse_tcp
msf6 exploit(linux/local/nested_namespace_idmap_limit_priv_esc) > back
```

```
msf6 > use exploit/linux/local/nested_namespace_idmap_limit_priv_esc:

Matching Modules
================

   #     Name                                                                               Disclosure Date  Rank       Check  Description
   -     ----                                                                               ---------------  ----       -----  -----------
   0     auxiliary/admin/2wire/xslt_password_reset                                          2007-08-15       normal     No     2Wire Cross-Site Request Forgery Password Reset Vulnerability
   1     auxiliary/admin/android/google_play_store_uxss_xframe_rce                                           normal     No     Android Browser RCE Through Google Play Store XFO

[...]

   4238  post/windows/wlan/wlan_disconnect                                                                   normal     No     Windows Disconnect Wireless Connection
   4239  post/windows/wlan/wlan_probe_request                                                                normal     No     Windows Send Probe Request Packets
   4240  post/windows/wlan/wlan_profile                                                                      normal     No     Windows Gather Wireless Profile


Interact with a module by name or index. For example info 4240, use 4240 or use post/windows/wlan/wlan_profile
```

This behavior can be reproduced with any value ending with `:` regardless of whether the previous text is a valid module path:

```
msf6 > use asdf:

Matching Modules
================

   #     Name                                                                               Disclosure Date  Rank       Check  Description
   -     ----                                                                               ---------------  ----       -----  -----------
   0     auxiliary/admin/2wire/xslt_password_reset                                          2007-08-15       normal     No     2Wire Cross-Site Request Forgery Password Reset Vulnerability
   1     auxiliary/admin/android/google_play_store_uxss_xframe_rce                                           normal     No     Android Browser RCE Through Google Play Store XFO

[...]

   4238  post/windows/wlan/wlan_disconnect                                                                   normal     No     Windows Disconnect Wireless Connection
   4239  post/windows/wlan/wlan_probe_request                                                                normal     No     Windows Send Probe Request Packets
   4240  post/windows/wlan/wlan_profile                                                                      normal     No     Windows Gather Wireless Profile


Interact with a module by name or index. For example info 4240, use 4240 or use post/windows/wlan/wlan_profile
```

```
msf6 > use asdf::
[-] Invalid argument(s)

Usage: search [<options>] [<keywords>:<value>]

Prepending a value with '-' will exclude any matching results.
If no options or keywords are provided, cached results are displayed.

OPTIONS:
  -h                Show this help information
  -o <file>         Send output to a file in csv format
  -S <string>       Regex pattern used to filter search results
  -u                Use module if there is one result

Keywords:
  aka         :  Modules with a matching AKA (also-known-as) name
  author      :  Modules written by this author
  arch        :  Modules affecting this architecture
  bid         :  Modules with a matching Bugtraq ID
  cve         :  Modules with a matching CVE ID
  edb         :  Modules with a matching Exploit-DB ID
  check       :  Modules that support the 'check' method
  date        :  Modules with a matching disclosure date
  description :  Modules with a matching description
  fullname    :  Modules with a matching full name
  mod_time    :  Modules with a matching modification date
  name        :  Modules with a matching descriptive name
  path        :  Modules with a matching path
  platform    :  Modules affecting this platform
  port        :  Modules with a matching port
  rank        :  Modules with a matching rank (Can be descriptive (ex: 'good') or numeric with comparison operators (ex: 'gte400'))
  ref         :  Modules with a matching ref
  reference   :  Modules with a matching reference
  target      :  Modules affecting this target
  type        :  Modules of a specific type (exploit, payload, auxiliary, encoder, evasion, post, or nop)

Examples:
  search cve:2009 type:exploit
  search cve:2009 type:exploit platform:-linux

[-] Failed to load module: asdf::
msf6 > 
```

### PR DESCRIPTION
**Pull request by nkitan**
Fixes #14768 by removing trailing : along side spaces before and after.

### CODE REVIEW

**Comment by gwillcox-r7** [comment]
Looks like this PR is breaking RSpec checks. Might be good to validate the regex you are using here. In particular the following checks are failing at the moment:

```
rspec './spec/lib/msf/core/modules/metadata/search_spec.rb[1:2:16:1:4]' # Msf::Modules::Metadata::Search#find on a module with a #description of "metasploit pro console" it should behave like search_filter should accept a query containing 'console pro'
rspec ./spec/lib/msf/core/modules/metadata/search_spec.rb:42 # Msf::Modules::Metadata::Search#parse_search_string is expected to eq {"arch"=>[["x64"], []], "author"=>[["egypt"], []]}
rspec ./spec/lib/msf/core/modules/metadata/search_spec.rb:39 # Msf::Modules::Metadata::Search#parse_search_string is expected to eq {"text"=>[["postgres", "login"], []]}
rspec ./spec/lib/msf/core/modules/metadata/search_spec.rb:38 # Msf::Modules::Metadata::Search#parse_search_string is expected to eq {"os"=>[["osx", "windows"], []]}
```

I can take a look at this again further tomorrow but for now I'd focus on trying to fix those errors if you can. Its possible, but unlikely, that the specs may need to change, however my guess is given that none of the expected output contains the `:` character that we are trying to filter, that something with the regex or similar is altering the contents of `search_string` in a way that is unexpected.

**Comment by gwillcox-r7** [comment]
@nkitan Quick note I updated your description as you need the `Fixes #*issue number*` syntax to link to the associated issue. Your previous description was close but because you had an extra work after `Fixes` and before the issue number, it wasn't linking things appropriately.

**Comment by nkitan** [comment]
Sure, I'll get to work on it asap

**Comment by nkitan** [comment]
@gwillcox-r7 I've changed the regex with a better version and it seems to work and completes all checks. please have a look whenever possible.

File: lib/msf/core/modules/metadata/search.rb [hunk 1]
```diff
@@ -34,7 +34,7 @@ def self.parse_search_string(search_string)
     search_string ||= ''
     search_string += ' '
     
-    search_string = search_string.gsub(/((\s*?\:+)+$)|((\:*\s)+$)/,'')
+    search_string = search_string.gsub(/((\:+\s*)\z|(\\s*:+)\z|(\:\s*:+ *)\z)/,'')
```
Comment (gwillcox-r7) [inline lib/msf/core/modules/metadata/search.rb:37]: My only comment on this is that you seem to be using one more capture group that is necessary for this regex to work, particularly given that we aren't interested in using any of the capture groups later on in the code. 

Also you seem to have an extra backslash here, so it should be `\s*:+`. Right now the regex for the middle expression is set up to explicitly match on `\s` as a string, reoccuring 0 or more times, followed by an `:` one or more times.

Proposed change below:
```suggestion
    search_string = search_string.gsub(/(\:+\s*)\z|(\s*:+)\z|(\:\s*:+ *)\z/,'')
```

Comment (gwillcox-r7) [reply lib/msf/core/modules/metadata/search.rb:37]: Upon further review I'm also not sure what the purpose of the last capture group is? It doesn't seem to fit in with what was originally the issue aka the trailing `:` character. Mind explaining what the purpose of that was?

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]

File: lib/msf/core/modules/metadata/search.rb (hunk 1 above)
Comment (gwillcox-r7) [reply lib/msf/core/modules/metadata/search.rb:37]: Alright yeah so even with that update to the regex you did, your third capture group will still fail to capture things like `use exploit/linux/local/nested_namespace_idmap_limit_priv_e:                                      :                 :`. Seems the regex needs a bit of updating.

Comment (gwillcox-r7) [reply lib/msf/core/modules/metadata/search.rb:37]: Ok found a solution, seems like `(\:+(\s*\:*)+\z)|(\s+(\s*\:*)+\z)` should resolve this. Here are some tests for reference:

```
msf6 > use exploit/linux/local/nested_namespace_idmap_limit_priv_e:::: :
"exploit/linux/local/nested_namespace_idmap_limit_priv_e:::: :  "
[*] Using configured payload linux/x86/meterpreter/reverse_tcp

Matching Modules
================

   #  Name                                                       Disclosure Date  Rank   Check  Description
   -  ----                                                       ---------------  ----   -----  -----------
   0  exploit/linux/local/nested_namespace_idmap_limit_priv_esc  2018-11-15       great  Yes    Linux Nested User Namespace idmap Limit Local Privilege Escalation


Interact with a module by name or index. For example info 0, use 0 or use exploit/linux/local/nested_namespace_idmap_limit_priv_esc

[*] Using exploit/linux/local/nested_namespace_idmap_limit_priv_esc
msf6 exploit(linux/local/nested_namespace_idmap_limit_priv_esc) > use exploit/linux/local/nested_namespace_idmap_limit_priv_e                                                   :
"exploit/linux/local/nested_namespace_idmap_limit_priv_e :  "
[*] Using configured payload linux/x86/meterpreter/reverse_tcp

Matching Modules
================

   #  Name                                                       Disclosure Date  Rank   Check  Description
   -  ----                                                       ---------------  ----   -----  -----------
   0  exploit/linux/local/nested_namespace_idmap_limit_priv_esc  2018-11-15       great  Yes    Linux Nested User Namespace idmap Limit Local Privilege Escalation


Interact with a module by name or index. For example info 0, use 0 or use exploit/linux/local/nested_namespace_idmap_limit_priv_esc

[*] Using exploit/linux/local/nested_namespace_idmap_limit_priv_esc
msf6 exploit(linux/local/nested_namespace_idmap_limit_priv_esc) > use exploit/linux/local/nested_namespace_idmap_limit_priv_e  :                  :     ::::                 :
"exploit/linux/local/nested_namespace_idmap_limit_priv_e : : :::: :  "
[*] Using configured payload linux/x86/meterpreter/reverse_tcp

Matching Modules
================

   #  Name                                                       Disclosure Date  Rank   Check  Description
   -  ----                                                       ---------------  ----   -----  -----------
   0  exploit/linux/local/nested_namespace_idmap_limit_priv_esc  2018-11-15       great  Yes    Linux Nested User Namespace idmap Limit Local Privilege Escalation


Interact with a module by name or index. For example info 0, use 0 or use exploit/linux/local/nested_namespace_idmap_limit_priv_esc

[*] Using exploit/linux/local/nested_namespace_idmap_limit_priv_esc
msf6 exploit(linux/local/nested_namespace_idmap_limit_priv_esc) > 
```

**Comment by gwillcox-r7** [comment]
Gah I'm really sorry about this @nkitan but I'm going to have to close this issue out as I can't merge this issue from your `master` branch. If you want to change the regex string to `(\:+(\s*\:*)+\z)|(\s+(\s*\:*)+\z)` as per my comment at https://github.com/rapid7/metasploit-framework/pull/14797#discussion_r582205714 and resubmit the PR from a unique branch, then ping me once this is done, I'll be happy to merge that in.

**Pull request closed without merging**